simulation/world.py

```diff
@@ -234,19 +239,52 @@ def _target_check(self) -> bool:
 
             # win type 1: the intruder has been in the target area for 3 seconds
             if (intruder.ticks_in_target * self.TIME_PER_TICK) >= 3.0:
-                intruder.reached_target = True
                 intruder.on_reached_target()
 
             # win type 2: the intruder has visited the target area twice with at least 3 seconds inbetween
             elif intruder.times_visited_target >= 2.0:
-                intruder.reached_target = True
+                intruder.on_reached_target()
 
-        # check if all intruders have reached the target
-        return all((intruder.reached_target for ID, intruder in self.intruders.items()))
+        # check if any intruders has reached the target
+        return any((intruder.reached_target for ID, intruder in self.intruders.items()))
 
     def setup(self):
+        patrolling_areas = self.create_patrolling_areas()
+        idx_pa = 0
+        idx_st = 0
+
         for ID, agent in self.agents.items():
             agent.setup(world=self)
+            if agent.type == 'PatrollingGuard':
```

Comment:
I think you should be able to move both of these calls to their respective `on_setup` methods? And that'd make sure that there's no additional logic here

Reply:
yes, i guess that should be possible. i just need to distribute the patrolling areas and towers between the agents, which i think is easier to do centrally than for each agent on its own